bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -195,8 +203,12 @@ private void sendAppendRowsRequest() throws IOException {
 
     ApiFuture<AppendRowsResponse> appendRowsResponseApiFuture =
         streamWriter.append(protoRows.build(), offset);
-    validateAppendRowsResponse(appendRowsResponseApiFuture, offset);
-
+    ApiFuture<AppendRowsResponse> validatedAppendRowsFuture =
+        ApiFutures.transformAsync(
+            appendRowsResponseApiFuture,
+            appendRowsResponse -> validateAppendRowsResponse(appendRowsResponse, offset),
```

Comment:
Is it ok if there is a race b/w rows? For e.g. a row1 is inserted by client at t1 and row2 inserted at t2 (t2 > t1), but row2 got inserted in the table before row1.

Reply:
Such a race should not occur. This is because the requests are being inserted on a single thread. With bidi streaming we are guaranteed that requests will be sent in order, and corresponding responses will arrive in order.